Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -120,7 +132,10 @@ - (NSString *)documentID {
 @dynamic metadata;
 
 - (FIRSnapshotMetadata *)metadata {
-  return _snapshot.GetMetadata();
+  if (!_cachedMetadata) {
+    _cachedMetadata = [[FIRSnapshotMetadata alloc] initWithMetadata:_snapshot.metadata()];
```

Comment:
Is writing a pointer to an ivar atomic?  Or is there a theoretical race condition where a thread could see half the pointer or something?

Reply:
Huh.  This is news to me!  I've specifically told developers in the past that RTDB's DataSnapshot is completely thread-safe and can be used across threads.  I thought the same was true for Firestore.  I always considered our entire API to be 100% thread-safe, and this is (part of) why we went to the trouble of having our single-threaded AsyncQueue and use immutable data structures everywhere, etc.

While I don't think this needs to be fixed in this (already-merged 😛) PR, I do think declaring our API _not_ thread-safe is a big concession.  We're super close to being thread-safe (I thought we already were) and it's not hard to make it actually be thread-safe.  So we should just fix it.

Reply:
OK. Reading through the literature on this point I think this is actually safe. I haven't been able to find a concise reference on this point, but [people believe it](https://forums.developer.apple.com/thread/20021#64167).

If this were a C/C++ pointer assignment, this would be undefined behavior. However, this is an Objective-C object pointer managed by ARC, and ARC guarantees that pointer assignments are atomic.

Note that this only works because every thread doing this will compute the same value, and if two threads race through here it's fine if both of them make a new metadata object.

Reply:
Yay!  Thanks for digging into it.  My faith in our SDKs is restored! 😄